librdmacm/rsocket.c

```diff
@@ -2983,19 +2987,123 @@ static uint64_t rs_time_us(void)
 	return now.tv_sec * 1000000 + now.tv_usec;
 }
 
+/* When mapping rpoll to poll, the events reported on the RDMA
+ * fd are independent from the events rpoll may be looking for.
+ * To avoid threads hanging in poll, whenever any event occurs,
+ * we need to wakeup all threads in poll, so that they can check
+ * if there has been a change on the rsockets they are monitoring.
+ * To support this, we 'gate' threads entering and leaving rpoll.
+ */
+static int rs_pollinit(void)
+{
+	int ret = 0;
+
+	pthread_mutex_lock(&mut);
+	if (pollsignal[0] || pollsignal[1])
+		goto unlock;
+
+	ret = socketpair(AF_UNIX, SOCK_STREAM, 0, pollsignal);
+	if (ret)
+		goto unlock;
+
+	/* Avoid hangs clearing the signal (reading fd) */
+	ret = set_fd_nonblock(pollsignal[0], true);
+	if (ret) {
+		close(pollsignal[0]);
+		close(pollsignal[1]);
+		pollsignal[0] = 0;
+		pollsignal[1] = 0;
+	}
+unlock:
+	pthread_mutex_unlock(&mut);
+	return ret;
+}
+
+/* When an event occurs, we must wait until the state of all rsockets
+ * has settled.  Then we need to re-check the rsocket state prior to
+ * blocking on poll().
+ */
+static int rs_poll_enter(void)
+{
+	pthread_mutex_lock(&mut);
+	if (suspendpoll) {
```

Comment:
Why use suspend poll rather than a trylock yeild loop here?  Does rs_poll_arm need to be called each time?

Reply:
We're not waiting on a lock here, but yes, we need to call rs_poll_arm() immediately before calling poll() to ensure that we notice all state changes on the rsockets that the thread is monitoring before blocking.  And that any state change after we block will generate an event on the fd.  There are threads currently in the poll() call that we need to wake-up and exit before we can proceed.

In other terms, we need all threads to *enter* poll together.  So threads currently in poll must exit and come get in line with threads waiting here.